vembrane/modules/annotate.py

```diff
@@ -0,0 +1,142 @@
+from typing import Iterator
+
+import numpy as np
+import pandas as pd
+import yaml
+from pysam.libcbcf import VariantFile, VariantRecord
+
+from ..common import check_expression
+from ..representations import Environment
+
+
+def add_subcommmand(subparsers):
+    parser = subparsers.add_parser("annotate")
+    parser.add_argument(
+        "config",
+        type=check_expression,
+        help="The configuration file.",
+    )
+    parser.add_argument(
+        "vcf", help="The file containing the variants.", nargs="?", default="-"
+    )
+    parser.add_argument(
+        "--output",
+        "-o",
+        default="-",
+        help="Output file, if not specified, output is written to STDOUT.",
+    )
+    parser.add_argument(
+        "--output-fmt",
+        "-O",
+        default="vcf",
+        choices=["vcf", "bcf", "uncompressed-bcf"],
+        help="Output format.",
+    )
+
+
+def annotate_vcf(
+    vcf: VariantFile,
+    expression: str,
+    ann_key: str,
+    ann_data: dict,
+    config: dict,
+) -> Iterator[tuple]:
+    env = Environment(expression, ann_key, vcf.header)
+
+    current_chrom = None
+    current_index = None
+    current_data = None
+    indices = []
+
+    record: VariantRecord
+    for idx, record in enumerate(vcf):
+        if not current_chrom == record.chrom:
+            current_chrom = record.chrom
+            chrom = None
+
+            # find the correct chrom name
+            tmp = current_chrom
+            if tmp.lower().startswith("chr"):
+                tmp = tmp[3:]
+            for prefix in ["", "chr", "Chr", "CHR"]:
+                if prefix + tmp in ann_data:
+                    chrom = prefix + tmp
+                    current_data = ann_data[chrom].to_records()
+                    current_index = 0
+                    indices = []
+        
+        if chrom:
+            # append possible intervals
+            while current_index < len(current_data) and (
+                current_data[current_index]["chromStart"] < record.start
+                indices.append(current_index)
+                current_index += 1
+
+            # copy only overlapping intervals
+            valid_indices = []
+            for index in indices:
+                if current_data[index]["chromEnd"] > record.start:
+                    valid_indices.append(index)
+
+            indices = valid_indices
+            if len(indices):
+                env.update_data(current_data[(np.array(indices))])
+                env.update_from_record(idx, record)
+                ann_values = env.table()
+
+                for name, value in zip(
+                    map(lambda x: x["value"]["vcf_name"], config["annotation"]["values"]),
+                    ann_values,
+                ):
+                    record.info[name] = float(value)
```

Comment:
We need to consider two things here:

1. `value` is not always meant to be a float, but depends on the "type" specified in the config file
2. multiple values must not overwrite existing values, but rather be appended to some collection, possibly depending on the "number" specified in the config file